docs/design/2021-02-09-ttl-table.md

```diff
@@ -0,0 +1,45 @@
+# Proposal: Time to live (TTL) Table
+
+- Author(s):     [Xiaoguang Sun](https://github.com/sunxiaoguang)
+- Last updated:  2021-02-09
+- Discussion at: https://github.com/pingcap/tidb/issues/22762
+
+## Abstract
+TTL Table makes it possible to automatically garbage collect data by user defined retention policy at either fine or coarse granularity.
+
+## Background
+There are application scenarios that data is only valuable for a certain period of time after ingestion and can be deleted permanently after expiration. Tracing, audit logs or push notification with expiration are examples of such applications. With the help of TTL tables, users can be relieved from taking care of the data life cycle in such tables. Therefore people are more willing to use TiDB as a general storage for such scenarios.
+
+## Proposal
+Introduce new `TTL` and `TTL_GRANULARITY` table options when creating or altering tables. By specifying `TTL`, users can expect expired data to be removed automatically. Additionally, users can also choose from `ROW` or `PARTITION` for `TTL_GRANULARITY` option to trade off collect granularity and cost to run such garbage collection. `ROW` mode evaluates expiry time against each row of data and reclaim space at row basis. This gives us the finest granularity and allows the most accurate expiration timing. `PARTITION` mode on the other hand partitions data according to its last update time. A background timer is going to rollover partitions and truncate all expired data within the oldest partition all at once.
```

Comment:
Can you show a full `SHOW CREATE TABLE` example so it becomes more clear? I assume this requires a timestamp to be collected, will this appear in the `SHOW CREATE TABLE`? It could be useful to see which rows are about to be purged.

Reply:
It could be something like:

```
CREATE TABLE ttl_table {
    id varchar(255),
    author varchar(255), 
    content varchar(65535),
    PRIMARY_KEY(id)  
} TTL=‘10m’ TTL_GRANULARITY=‘ROW’;
```

```
CREATE TABLE ttl_table {
    id varchar(255),
    author varchar(255), 
    content varchar(65535),
    PRIMARY_KEY(id)  
} TTL=‘10m’ TTL_GRANULARITY=‘PARTITION’;
```

Reply:
Oh I see! So it uses partitioning internally, but not partition syntax. Is there any restriction such as TTL tables can not be partitioned?

Reply:
Hm, as far as I know there is no restriction when granularity is set to partition as long as itself is not a partitioned table. If people want to use row granularity, it works for partition table as well.

Reply:
Sorry, that is what I meant with my question. Please document that Partitioned tables do not support `TTL_GRANULARITY=‘PARTITION’`.

Reply:
Oh I see what you mean know. It is possible to archive this by manually managing and truncating partitioned table, but it's error prone and tedious. I guess nobody would like to do that if it can be done automatically otherwise.

Reply:
> It could be something like:

```
CREATE TABLE ttl_table {
    id varchar(255),
    author varchar(255), 
    content varchar(65535),
    PRIMARY_KEY(id)  
} TTL=‘10m’ TTL_GRANULARITY=‘ROW’;
CREATE TABLE ttl_table {
    id varchar(255),
    author varchar(255), 
    content varchar(65535),
    PRIMARY_KEY(id)  
} TTL=‘10m’ TTL_GRANULARITY=‘PARTITION’;
```

1. Could you write the syntax to this doc?
2. If we consider the compatibility with MySQL(say for example, need to replicate the table to MySQL, maybe it's necessary to support the ['TiDB-specific commet style'](https://docs.pingcap.com/tidb/stable/comment-syntax#tidb-specific-comment-syntax) so it won't be broken in MySQL).

Reply:
(Just a reminder)I noticed that [MyRocks specified the TTL ](https://www.programmersought.com/article/34771345021/)in a different way like:

```
CREATE TABLE t1 (
  a bigint(20) NOT NULL,
  b int NOT NULL,
  ts bigint(20) UNSIGNED NOT NULL,
  PRIMARY KEY (a),
  KEY kb (b)
) ENGINE=rocksdb
COMMENT='ttl_duration=1;ttl_col=ts;'
```

However, I believe that this is caused by the fact that MyRocks is only one of the storage engines for MySQL, so 'TTL' cannot be specified as a 'table option', so I would say it is better to keep it as a table option, as what you are suggesting now.